src/java.base/share/classes/java/util/stream/Gatherers.java

```diff
@@ -0,0 +1,657 @@
+        Objects.requireNonNull(initial, "'initial' must not be null");
+        Objects.requireNonNull(scanner, "'scanner' must not be null");
+
+        class State {
+            R current = initial.get();
+            boolean integrate(T element, Downstream<? super R> downstream) {
+                return downstream.push(current = scanner.apply(current, element));
+            }
+        }
+
+        return Gatherer.ofSequential(State::new,
+                                     Integrator.<State,T, R>ofGreedy(State::integrate));
+    }
+
+    /**
+     * An operation which executes operations concurrently
+     * with a fixed window of max concurrency, using VirtualThreads.
+     * This operation preserves the ordering of the stream.
+     *
+     * <p>In progress tasks will be attempted to be cancelled,
+     * on a best-effort basis, in situations where the downstream no longer
+     * wants to receive any more elements.
+     *
+     * <p>If the mapper throws an exception during evaluation of this Gatherer,
+     * and the result of that invocation is to be produced to the downstream,
+     * then that exception will instead be rethrown as a {@link RuntimeException}.
+     *
+     * @param maxConcurrency the maximum concurrency desired
+     * @param mapper a function to be executed concurrently
+     * @param <T> the type of input
+     * @param <R> the type of output
+     * @return a new Gatherer
+     * @throws IllegalArgumentException if maxConcurrency is less than 1
+     * @throws NullPointerException if mapper is null
+     */
+    public static <T, R> Gatherer<T,?,R> mapConcurrent(
+            final int maxConcurrency,
+            final Function<? super T, ? extends R> mapper) {
+        if (maxConcurrency <= 0)
+            throw new IllegalArgumentException(
+                    "'maxConcurrency' needs to be greater than 0");
+
+        Objects.requireNonNull(mapper, "'mapper' must not be null");
+
+        class State {
+            final ArrayDeque<Future<R>> window = new ArrayDeque<>(maxConcurrency);
+            final Semaphore windowLock = new Semaphore(maxConcurrency);
+
+            final boolean integrate(T element,
+                                    Downstream<? super R> downstream) {
+                if (!downstream.isRejecting())
+                    createTaskFor(element);
+                return flush(0, downstream);
+            }
+
+            final void createTaskFor(T element) {
+                windowLock.acquireUninterruptibly();
+
+                var task = new FutureTask<R>(() -> {
+                    try {
+                        return mapper.apply(element);
+                    } finally {
+                        windowLock.release();
+                    }
+                });
+
+                var wasAddedToWindow = window.add(task);
+                assert wasAddedToWindow;
+
+                Thread.startVirtualThread(task);
```

Comment:
At some point it might be worth seeing this should should use Executors.newVirtualThreadPerTaskExecutor(), that would at last simplify the cleanup as it would be executor.shutdownNow.  The serviceability benefit would be that an observe on a far hill taking a thread dump would have the virtual threads grouped for each mapConcurrent gatherer.

Reply:
Yeah, that's a good point. I'll try that after this PR. :+1: